src/electron/ipc/ipc-renderer-shim.ts

```diff
@@ -0,0 +1,74 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+import { Action } from 'common/flux/action';
+import { IpcRenderer } from 'electron';
+import { SetSizePayload } from 'electron/flux/action/window-frame-actions-payloads';
+import {
+    IPC_APP_VERSION_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_CLOSE_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_ENTERFULLSCREEN_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_MAXIMIZE_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_MINIMIZE_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_RESTORE_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_SETSIZEANDCENTER_CHANNEL_NAME,
+    IPC_BROWSERWINDOW_UNMAXIMIZE_CHANNEL_NAME,
+    IPC_MAIN_WINDOW_INITIALIZED_CHANNEL_NAME,
+} from 'electron/ipc/ipc-channel-names';
+
+export class IpcRendererShim {
+    public constructor(private readonly ipcRenderer: IpcRenderer) {}
+
+    private readonly invokeScope: string = 'IpcRendererShim';
+
+    public initialize(): void {
+        this.ipcRenderer.on(IPC_BROWSERWINDOW_MAXIMIZE_CHANNEL_NAME, this.onMaximize);
+        this.ipcRenderer.on(IPC_BROWSERWINDOW_UNMAXIMIZE_CHANNEL_NAME, this.onUnmaximize);
+        this.ipcRenderer.on(IPC_BROWSERWINDOW_ENTERFULLSCREEN_CHANNEL_NAME, this.onEnterFullScreen);
+    }
+
+    private onMaximize = (): void => {
+        this.maximizeEvent.invoke(null, this.invokeScope);
+    };
+
+    private onEnterFullScreen = (): void => {
+        this.enterFullScreenEvent.invoke(null, this.invokeScope);
+    };
+
+    private onUnmaximize = (): void => {
+        this.unmaximizeEvent.invoke(null, this.invokeScope);
+    };
+
+    // Listen to these events to receive data sent TO renderer process
+    public readonly maximizeEvent = new Action<void>();
+    public readonly unmaximizeEvent = new Action<void>();
+    public readonly enterFullScreenEvent = new Action<void>();
+
+    // Call these methods to send data FROM renderer process
+    public initializeWindow = (): void => {
+        this.ipcRenderer.send(IPC_MAIN_WINDOW_INITIALIZED_CHANNEL_NAME);
+    };
+
+    public getVersion = (): string => {
+        return this.ipcRenderer.sendSync(IPC_APP_VERSION_CHANNEL_NAME);
```

Comment:
It doesn't seem like this is worse than remote, but I wonder if we're able to use invoke (async) instead - or whether we should treat it similar to `getPersistedData` in `renderer-initializer.ts`. There's a warning [here](https://www.electronjs.org/docs/api/ipc-renderer#ipcrenderersendsyncchannel-args) about `sendSync` blocking the renderer process like the remote module did. In this case the async signature would bubble up to its usage in `scan-controller.ts`.

Reply:
I'm not a fan of the sync option, but it was the "least bad" option I could find. Making the call async means changing the signature of the `AppDataAdapter` interface, and that is used in lots of places, including many that are unrelated to Electron. It's not persisted data, either, since it's a function of the bits instead of the user store. I had hoped to make it async but to find a way to block the calling thread (like .NET's `Thread.Wait` method), but that concept isn't supported in JavaScript.

Reply:
(chatted offline) seems like options are:
- move version into InsightsConfiguration or a store since it's a constant value
- make this method async and update all its usages

It'd be nice to explore more but doesn't need to be part of this PR given we already were sync here